lib/internal/modules/esm/loader.js

```diff
@@ -194,13 +218,63 @@ class ESMLoader {
         this.#loaders,
         FunctionPrototypeBind(loader, null), // [1]
       );
+      if (getFileSystem) ArrayPrototypePush(
+        this.#fileSystemBuilders,
+        FunctionPrototypeBind(getFileSystem, null), // [1]
+      );
     }
 
     // [1] ensure hook function is not bound to ESMLoader instance
 
+    this.buildFileSystem();
     this.preload();
   }
 
+  buildFileSystem() {
+    // Note: makes assumptions as to how chaining will work to demonstrate
+    // the capability; subject to change once chaining's API is finalized.
+    const fileSystemFactories = [...this.#fileSystemBuilders];
+
+    const defaultFileSystemFactory = fileSystemFactories[0];
+    let finalFileSystem =
+      defaultFileSystemFactory();
+
+    const asyncKeys = ArrayPrototypeFilter(
+      ObjectKeys(finalFileSystem),
+      (name) => !StringPrototypeEndsWith(name, 'Sync'),
+    );
+
+    for (let i = 1; i < fileSystemFactories.length; ++i) {
+      const currentFileSystem = finalFileSystem;
+      const fileSystem = fileSystemFactories[i](() => currentFileSystem);
```

Comment:
Mostly to map the signature of other hooks, which provide the next hook function rather than their results (granted, they accept inputs whereas this one doesn't, so it *could* be possible, but by consistency I feel more comfortable keeping the same pattern).

Reply:
eventually this should do a copy operation rather than direct reference to avoid people purposefully doing mutation becoming breaking change worries.